src/cache.rs

```diff
@@ -270,33 +422,40 @@ impl<T: Hash + PartialEq> Constraint<T> {
     where
         F: FnMut(&T) -> u128,
     {
-        let calls = self.0.borrow();
-        ACCELERATOR.with(|accelerator| {
-            let mut map = accelerator.borrow_mut();
-            calls.iter().all(|entry| {
-                *map.entry((id, entry.both)).or_insert_with(|| f(&entry.args))
-                    == entry.ret
+        let accelerator = accelerator(id);
+        let inner = self.0.read();
+        if let Some(accelerator) = accelerator {
+            let mut map = accelerator.lock();
+            inner.values().all(|entry| {
+                *map.entry(entry.both).or_insert_with(|| f(&entry.args)) == entry.ret
             })
-        })
+        } else {
+            inner.values().all(|entry| f(&entry.args) == entry.ret)
+        }
     }
 
     /// Replay all input-output pairs.
     #[inline]
-    pub fn replay<F>(&self, mut f: F)
+    pub fn replay<F>(&self, _: F)
     where
         F: FnMut(&T),
     {
-        for entry in self.0.borrow().iter() {
-            if entry.mutable {
-                f(&entry.args);
-            }
+        #[cfg(debug_assertions)]
+        for entry in self.0.read().values() {
+            assert!(!entry.mutable);
         }
```

Comment:
I think the mutable state should be irrepresentable.

The `Call` struct and its `args` field are named in a bit misleading way. The `args` are called `Call` in the macro and encapsulate not only the arguments, but also which tracked methods was called (by being an enum). This enum technically also statically determines whether the call was mutable or not, so the mutable bool could be completely removed if the method was bound on a trait that says whether the call was mutable.

I think a nicer design could be something like this:

```rust
struct ImmutableConstraint<T: Call>(RwLock<HashMap<u128, ConstraintEntry<T>>>);

struct MutableConstraint<T: Call>(RwLock<Vec<ConstraintEntry<T>>>);

struct ConstraintEntry<T: Call> {
  call: T,
  call_hash: u128,
  ret_hash: u128,
}

trait Call {
  fn is_mutable(&self) -> bool;
}
```